[BUG] Metadata serialization.

**Describe the bug**
Metadata object not being serialized correctly. The items within `SnapshotMetadata.caller_info.args` should be serialized like any other object being serialized. 

**To Reproduce**
```python
from pytest import fixture

class Foo:
    pass

@fixture
def fxtr()
    return Foo()

def test_function(fxtr, snapshot):
    snapshot.assert_match(True)
```

**Expected behavior**
This should successfully snapshot the value `True`

**Screenshots**
Output from original error:
```
self = <snappiershot.serializers.json.JsonSerializer object at 0x11b8d1160>, value = <FlaskClient <Flask 'ace_farm.manager'>>

    def default(self, value: Any) -> Any:
        """ Encode a value into a serializable object.
    
        This method only gets called when the value is not naturally-serializable.
          (Naturally serializable objects are booleans, floats, strings, etc.)
          See: https://docs.python.org/3/library/json.html#py-to-json-table
    
        Any custom encoders must return a dictionary. This will allow them to
          be deserialized by the `snappiershot.serializers.json.JsonDeserializer`
          using object hooks.
    
        Encoding of collections (sets and tuples) are done in a pre-processing step
          within the ``JsonSerializer.encode`` method. That method should always be
          called prior to this method.
    
        Args:
            value: The python object to encode.
        """
        if isinstance(value, Number):
            return self.encode_numeric(value)
    
        if isinstance(value, DATETIME_TYPES):
            return self.encode_datetime(value)
    
>       raise NotImplementedError(  # pragma: no cover
            f"Encoding for this object is not yet implemented: {value} ({type(value)})"
        )
E       NotImplementedError: Encoding for this object is not yet implemented: <FlaskClient <Flask 'ace_farm.manager'>> (<class 'flask.testing.FlaskClient'>)

environments/ace/.venv/lib/python3.8/site-packages/snappiershot/serializers/json.py:79: NotImplementedError
```

**Versions:**
 - Snappiershot version: 0.2.0

**Additional context**
This should be fixable by running the `SnapshotMetadata.caller_info.args` through the `default_encode_value` utility function. Additionally we may want to restrict what fixtures are snapshotted, possible through `__module__` checks.

--- a/snappiershot/snapshot/_file.py
+++ b/snappiershot/snapshot/_file.py
@@ -1,11 +1,11 @@
 """ Interface for the snapshot files. """
 from pathlib import Path
-from typing import Any, Dict, List, Union
+from typing import Any, Dict, List, Union, cast
 
 from ..config import Config
 from ..constants import SnapshotKeys
 from ..serializers.io import parse_snapshot_file, write_json_file
-from ..serializers.utils import get_snapshot_file
+from ..serializers.utils import default_encode_value, get_snapshot_file
 from .metadata import SnapshotMetadata
 from .status import SnapshotStatus
 
@@ -187,8 +187,9 @@ class _SnapshotFile:
                 return function_snapshot[SnapshotKeys.snapshots]
 
         # If no sub-section exists, create it.
+        encoded_metadata = cast(dict, default_encode_value(metadata.as_dict()))
         function_snapshots.append(
-            dict(metadata={**metadata.as_dict(), "update_on_next_run": False}, snapshots=[])
+            dict(metadata={**encoded_metadata, "update_on_next_run": False}, snapshots=[])
         )
         return function_snapshots[-1][SnapshotKeys.snapshots]
 

--- a/tests/test_snapshot/test_file.py
+++ b/tests/test_snapshot/test_file.py
@@ -1,10 +1,12 @@
 """ Tests for snappiershot/snapshot/_file.py """
 import json
 from pathlib import Path
+from types import SimpleNamespace
 from typing import Dict, List
 
 import pytest
 from snappiershot.config import Config
+from snappiershot.inspection import CallerInfo
 from snappiershot.snapshot._file import _NO_SNAPSHOT, _SnapshotFile
 from snappiershot.snapshot.metadata import SnapshotMetadata
 from snappiershot.snapshot.status import SnapshotStatus
@@ -220,3 +222,20 @@ class TestSnapshotFile:
         assert snapshot_file.exists() == file_written
         if file_written:
             assert snapshot_file_object._snapshot_statuses == expected_statuses
+
+    @staticmethod
+    def test_encoded_metadata(config: Config, snapshot_file: Path):
+        """ Test that SnapshotMetadata gets encoded to eliminate serialization errors. """
+        # Arrange
+        args = dict(unhandled_type=SimpleNamespace(complex=3 + 4j))
+        caller_info = CallerInfo(snapshot_file, "test_function", args)
+        metadata = SnapshotMetadata(caller_info, update_on_next_run=False)
+        snapshot_file_object = _SnapshotFile(config, metadata)
+
+        # Act
+        snapshot_file_object._changed_flag = True
+        snapshot_file_object.write()
+
+        # Assert
+        assert snapshot_file.exists()
+        print(snapshot_file.read_text())

